internal/docs/design.md

```diff
@@ -184,6 +187,102 @@ needed). However, this would result in inconsistent names over time (e.g., some
 packages would expose `New` with an `Options`, while others would expose
 `NewWithOptions`).
 
+### URLs
+
+To enable the [Backing services factor][] of a Twelve-Factor Application, Go
+Cloud includes the ability to construct each of its API objects using
+identifying URLs. The concrete type should have API like this:
+
+```go
+package foo
+
+// A type that implements WidgetURLOpener can open widgets based on a URL.
+// The opener must not modify the URL argument. OpenWidgetURL must be safe to
+// call from multiple goroutines.
+//
+// For maximum flexibility, WidgetURLOpeners should not assume that the URL
+// has a particular scheme.
+type WidgetURLOpener interface {
+  OpenWidgetURL(ctx context.Context, u *url.URL) (*Widget, error)
+}
+
+// A type that implements NamedWidgetURLOpener has a preferred scheme for
+// the URLs it opens. WidgetURLScheme must be safe to call from multiple
+// goroutines.
+type NamedWidgetURLOpener interface {
```

Comment:
Why do you need the extra complexity of the preferred scheme (basically, this type, and AddWidget method)?

Reply:
I'm not sure I get the need for this extra interface this either. E.g., why not have each provider impl expose its default scheme(s) explicitly? The code setting up a `URLMux` can use them or not, so there's no loss of control. It might help me to understand when you add an example of what a provider impl would look like.

Reply:
I am envisioning something like this (names are fungible):

*  the concrete type exposes a way to register an `URLOpener` to a scheme -- your `RegisterWidget` -- and nothing else about schemes/naming.
*  each provider impl has a public function `RegisterOpener` that takes a mux and an `URLOpener` and registers the opener for the default scheme(s).
*  each provider impl has a package `init` that calls its `RegisterOpener` with the default mux and provider-default `URLOpener`.

Each driver impl can define its default schemes and default Opener behavior to be exposed via the default mux. Users can choose to create their own mux, create openers with custom settings (e.g., creds) but still register for the default schemes (by calling `provider.Register(mymux, myopener)`, or choose their own schemes (by calling `foo.Register(scheme, opener)` directly).

Reply:
@vangent How necessary is this given that we now have a global registry? There's enough going on in this design as is: I'm not seeing your solution as saving the user much effort as just registering the openers with the schemes explicitly and it loses a lot of the clarity.

Reply:
IIUC, your original proposal had each provider impl defining a `NamedWidgetURLOpener` per scheme. This is replacing those structs-implementing-interface with a single function, which seems simpler to me.

We could also drop this idea entirely, and require users that don't want the default/global mux to create a new mux and explicitly register specific schemes for it. We'd have to expose the default scheme as a public constant, because most users won't have any reason to change the scheme. I guess I'm OK with that; IMHO my proposal is clearly better if there are multiple schemes (e.g., aliases), but maybe we just don't do that.